File: backend/src/app/models/task.py
from datetime import datetime
from typing import Optional, Dict, Any
from sqlalchemy import Column, Integer, String, Text, JSON, DateTime, Boolean, ForeignKey
from sqlalchemy.dialects.mysql import TINYINT
from .base import BaseModel

class Task(BaseModel):
    __tablename__ = "tasks"
    
    user_id = Column(Integer, nullable=True, comment="用户ID")
    workflow_type = Column(String(32), nullable=False, comment="工作流类型")
    # 状态码约定：0待处理,1处理中,2成功,3失败,4跳过,5取消,6部分完成
    # 备注：将使用 6 作为 INTERRUPTED (中断) 状态的约定（用于 TaskStep/Scene 的中断语义）
    status = Column(TINYINT, nullable=False, default=0, comment="状态码：0待处理,1处理中,2成功,3失败,4跳过,5取消,6部分完成")
    params = Column(JSON, nullable=True, comment="任务参数JSON")
    result = Column(JSON, nullable=True, comment="任务结果JSON")
    error_msg = Column(String(255), nullable=True, comment="错误信息")
    is_deleted = Column(Boolean, nullable=False, default=False, comment="是否删除")
    
    # 新增字段
    task_config = Column(JSON, nullable=True, comment="任务完整配置JSON")
    # 自动/手动模式：'auto' | 'manual' （优先从单独列读取，兼容旧的 task_config.mode）
    mode = Column(String(16), nullable=False, default='auto', comment="任务执行模式：auto/manual")
    progress = Column(Integer, nullable=False, default=0, comment="任务进度 0-100")
    total_scenes = Column(Integer, nullable=True, comment="总分镜数")
    completed_scenes = Column(Integer, nullable=False, default=0, comment="已完成分镜数")
    providers = Column(JSON, nullable=True, comment="功能模块选用的服务提供商映射，例如 {'image':'runninghub'}")
    merged_video_url = Column(String(500), nullable=True, comment="分镜拼接后视频URL（未加背景音乐）")
    final_video_url = Column(String(500), nullable=True, comment="最终成片URL（完成配乐/滤镜等处理）")
    selected_voice_id = Column(String(128), nullable=True, comment="Fish Audio voice_id chosen for this task")
    selected_voice_name = Column(String(128), nullable=True, comment="Display name of the selected voice")
    subtitle_style_id = Column(Integer, ForeignKey("subtitle_styles.id", ondelete="SET NULL"), nullable=True, comment="选择的字幕样式 ID")
    subtitle_style_snapshot = Column(JSON, nullable=True, comment="字幕样式快照 JSON")

class TaskStep(BaseModel):
    __tablename__ = "task_steps"
    
    task_id = Column(Integer, ForeignKey("tasks.id"), nullable=False, comment="关联任务ID")
    step_name = Column(String(64), nullable=False, comment="步骤名称")
    seq = Column(Integer, nullable=False, comment="步骤顺序")
    # 状态码约定：0待处理,1处理中,2成功,3失败,4跳过,5取消,6中断
    status = Column(TINYINT, nullable=False, default=0, comment="状态码：0待处理,1处理中,2成功,3失败,4跳过,5取消,6中断")
    params = Column(JSON, nullable=True, comment="步骤参数JSON")
    result = Column(JSON, nullable=True, comment="步骤结果JSON")
    error_msg = Column(String(255), nullable=True, comment="错误信息")
    started_at = Column(DateTime, nullable=True, comment="开始时间")
    finished_at = Column(DateTime, nullable=True, comment="完成时间")
    
    # 新增字段
    retry_count = Column(Integer, nullable=False, default=0, comment="重试次数")
    max_retries = Column(Integer, nullable=False, default=3, comment="最大重试次数")
    progress = Column(Integer, nullable=False, default=0, comment="步骤进度 0-100")
    provider = Column(String(64), nullable=True, comment="本步骤使用的服务提供商")
    external_task_id = Column(String(128), nullable=True, comment="外部服务返回的任务ID或Job ID")
    context = Column(JSON, nullable=True, comment="步骤运行时上下文信息，例如排队状态、回调信息")

File: web/src/types/task.ts
import type { SubtitleDocument } from './subtitle'

export interface TaskParams {
  title?: string
  description?: string
  reference_video?: string
}

export interface TaskConfig {
  scene_count?: number
  language?: string
  provider?: string | null
  style_preset_id?: number | null
  subtitle_style_id?: number | null
  subtitle_style_snapshot?: Record<string, unknown> | null
  bgm_asset_id?: number | null
  providers?: Record<string, string> | null
  audio_trim_silence?: boolean
  [key: string]: unknown
}

export interface TaskSummary {
  id: number
  workflow_type: string
  status: number
  progress: number
  total_scenes: number | null
  completed_scenes: number
  params: TaskParams | null
  result: Record<string, unknown> | null
  task_config: TaskConfig | null
  providers: Record<string, string> | null
  error_msg: string | null
  created_at: string
  updated_at: string
  mode: 'auto' | 'manual'
  merged_video_url: string | null
  final_video_url: string | null
  selected_voice_id?: string | null
  selected_voice_name?: string | null
  subtitle_style_id?: number | null
  subtitle_style_snapshot?: Record<string, unknown> | null
}

export interface TaskStep {
  id: number
  task_id: number
  step_name: string
  seq: number
  status: number
  progress: number
  retry_count: number
  max_retries: number
  params: Record<string, unknown> | null
  result: Record<string, unknown> | null
  error_msg: string | null
  started_at: string | null
  finished_at: string | null
}

export interface SceneRecord {
  id: number
  task_id: number
  seq: number
  status: number
  narration_text: string | null
  narration_word_count: number | null
  image_prompt: string | null
  video_prompt: string | null
  image_status: number
  audio_status: number
  audio_duration: number | null
  video_status: number
  merge_status: number
  image_url: string | null
  audio_url: string | null
  raw_video_url: string | null
  merge_video_url: string | null
  merge_video_provider: string | null
  image_meta: Record<string, unknown> | null
  audio_meta: Record<string, unknown> | null
  video_meta: Record<string, unknown> | null
  merge_meta: Record<string, unknown> | null
  image_retry_count: number
  audio_retry_count: number
  video_retry_count: number
  merge_retry_count: number
  params: Record<string, unknown> | null
  result: Record<string, unknown> | null
  error_msg: string | null
  started_at: string | null
  finished_at: string | null
}

export interface TaskDetail {
  task: TaskSummary
  steps: TaskStep[]
  scenes: SceneRecord[]
  subtitle_document?: SubtitleDocument | null
}

export interface CreateTaskPayload {
  title: string
  description: string
  reference_video?: string
  mode: 'auto' | 'manual'
  scene_count: number
  language: string
  audio_voice_id: string
  audio_voice_service?: string
  provider?: string | null
  media_tool?: string | null
  style_preset_id?: number
  subtitle_style_id?: number | null
  bgm_asset_id?: number | null
  audio_trim_silence?: boolean
}

export interface StoryboardScriptScene {
  scene_number: number
  scene_name?: string | null
  name?: string | null
  visual: string
  animation?: string | null
  narration?: string | null
  dialogue?: string | null
  [key: string]: unknown
}

export interface StoryboardScriptPayload {
  script: StoryboardScriptScene[]
}

export interface StoryboardScriptImportResponse {
  task_id: number
  scene_count: number
  provider: string
  auto_mode: 'auto' | 'manual' | null
  auto_dispatched: boolean
}

export interface UpdateTaskPayload {
  title?: string
  description?: string
  reference_video?: string
  mode?: 'auto' | 'manual'
  scene_count?: number
  language?: string
  audio_voice_id?: string
  audio_voice_service?: string
  provider?: string | null
  media_tool?: string | null
  style_preset_id?: number | null
  subtitle_style_id?: number | null
  bgm_asset_id?: number | null
  audio_trim_silence?: boolean
}

export interface VoiceOption {
  id: number
  service_name: string
  option_type: string
  option_key: string
  option_value: string
  option_name: string | null
  description: string | null
  is_default: boolean
  is_active: boolean
  meta_data: Record<string, unknown> | null
}
